// ngx-script-loader/esm5/ngx-script-loader.js
import { Injectable, Component, Input, ElementRef, Output, EventEmitter, NgModule } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/operator/take';
import 'rxjs/add/operator/shareReplay';
import { CommonModule } from '@angular/common';

var ScriptService = /** @class */ (function () {
    function ScriptService() {
        this.scriptsLoaders = {};
    }
    ScriptService.prototype.loadScript = function (url, attributes, targetEl) {
        if (targetEl === void 0) { targetEl = 'head'; }
        if (url in this.scriptsLoaders) {
            return this.scriptsLoaders[url];
        }
        this.scriptsLoaders[url] = Observable.create(function (observer) {
            var script = document.createElement('script');
            if (attributes) {
                for (var key in attributes) {
                    if (attributes.hasOwnProperty(key)) {
                        script.setAttribute(key, attributes[key]);
                    }
                }
            }
            script.src = url;
            script.onload = function (event) {
                observer.next(event);
                observer.complete();
            };
            script.onerror = function (err) {
                observer.error(err);
            };
            var targetElement = typeof targetEl === 'string' ? document.querySelector(targetEl) : targetEl;
            targetElement.appendChild(script);
        })
            .take(1)
            .shareReplay(1);
        return this.scriptsLoaders[url];
    };
    return ScriptService;
}());
ScriptService.decorators = [
    { type: Injectable },
];
var ScriptComponent = /** @class */ (function () {
    function ScriptComponent(scriptService, element) {
        this.scriptService = scriptService;
        this.element = element;
        this.attributes = {};
        this.load = new EventEmitter();
        this.error = new EventEmitter();
    }
    ScriptComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.scriptService.loadScript(this.src, this.attributes, this.element.nativeElement)
            .subscribe(function (event) {
            _this.load.emit(event);
        }, function (error) {
            _this.error.emit(error);
        });
    };
    return ScriptComponent;
}());
ScriptComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-script',
                template: '',
                styles: []
            },] },
];
ScriptComponent.ctorParameters = function () { return [
    { type: ScriptService, },
    { type: ElementRef, },
]; };
ScriptComponent.propDecorators = {
    "src": [{ type: Input },],
    "attributes": [{ type: Input },],
    "load": [{ type: Output },],
    "error": [{ type: Output },],
};
var ScriptLoaderModule = /** @class */ (function () {
    function ScriptLoaderModule() {
    }
    ScriptLoaderModule.forRoot = function () {
        return {
            ngModule: ScriptLoaderModule,
            providers: [
                ScriptService
            ]
        };
    };
    return ScriptLoaderModule;
}());
ScriptLoaderModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    ScriptComponent
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    ScriptComponent
                ]
            },] },
];

export { ScriptService, ScriptComponent, ScriptLoaderModule };
